### ISSUE TITLE
No sound with dmplay, meanwhile dmrender converts the same file successfully.

### ISSUE THREAD

**Issue by mzso** [frabert/libdmusic#17]
Version 0.1.1. The same file that's successfully converted by dmrender, with an analogous command line doesn't produce any sound in dmplay.

**Comment by frabert** [comment]
Does dmplay show anything on the console after being invoked?

**Comment by mzso** [comment]
No. Just he usual:

`Begin rendering... Rendering started. Insert the next segment to be played, or 'exit' to end playback.`

**Comment by frabert** [comment]
I've tried adding some logging to dmplay, maybe it will show where the fault is

**Comment by mzso** [comment]
I still get nothing:
```
dmplay.exe -c 2 -s 48000 AMB1.SGT
Opening audio device...  done.
Loading segment... done.
Start playback...  done.
Begin rendering... Rendering started. Insert the next segment to be played, or 'exit' to end playback.
```

This works for dmrender:

```
dmrender.exe -c 2 -s 48000 -l 60 AMB1.SGT amb1.wav
Loading segment... done.
Start playback...  done.
Begin rendering...
Progress: 99%
Rendering done.
```

PS:
Adding the version number to the releases would be nice, so I wouldn't need to manually name them each time.

**Comment by frabert** [comment]
Thanks for the suggestion, I'll see what I can do.

Do you have more than one sound device in your system? Maybe your sound card has multiple 'virtual' devices for digital and analog signal or something like that...

**Comment by mzso** [comment]
[**@frabert**](https://github.com/frabert) commented on [2018. aug. 28. 15:21 CEST](https://github.com/frabert/libdmusic/issues/17#issuecomment-416581522 "2018-08-28T13:21:10Z - Replied by Github Reply Comments"):
> Thanks for the suggestion, I'll see what I can do.
> 
> Do you have more than one sound device in your system? Maybe your sound card has multiple 'virtual' devices for digital and analog signal or something like that...

One integrated audio adapter at the moment. The analog speaker is the default so that shouldn't be a problem unless somehow dmplay forces itself on the digital output. Though I don't see anything on windows' volume bar for that device, unlike the bar for the speaker if I play back the wav produced by dmrender.

**Comment by frabert** [comment]
I added a flag `-l` to dmplay to list all the available sound devices in the system, and `-d [deviceNum]` to choose the one to use. At the moment, it's the only thing I can think of that may cause the problem, unfortunately...

**Comment by mzso** [comment]
[**@frabert**](https://github.com/frabert) commented on [2018. aug. 29. 12:07 CEST](https://github.com/frabert/libdmusic/issues/17#issuecomment-416899406 "2018-08-29T10:07:17Z - Replied by Github Reply Comments"):
> I added a flag `-l` to dmplay to list all the available sound devices in the system, and `-d [deviceNum]` to choose the one to use. At the moment, it's the only thing I can think of that may cause the problem, unfortunately...

It looks like the device has nothing to do with it. It fails to produce any audio when the output is not set to stereo for my sound adapter.
(Strangely at first it seemed to work when I tried right now, but now it doesn't, only when I change to stereo. Unless I'm misremembering.)

(Also, why do I get an error when I don't match the sampling rate setting of the audio device? Normally that should be resampled automatically by the driver, shouldn't it?)

**Comment by mzso** [comment]
Update:
I can get sound if I change the channel number in dmplay to match the system settings. Although it produces a garbled abnormal output. (multi-channel is broken with dmrender too, though it's broken in a different sounding way)

It looks like outputting audio has a number of issues.

**Comment by frabert** [comment]
Interesting, indeed dmusic doesn't support multichannel (>2 channels) output... I can try making it output silence on any channel following the first two

**Comment by frabert** [comment]
Yeah RtAudio seems lazier that Portaudio, it doesn't like to convert between sampling rates / channel counts, but it is nicer to use from C++, I think it is staying for now

**Comment by mzso** [comment]
[**@frabert**](https://github.com/frabert) commented on [2018. aug. 29. 13:16 CEST](https://github.com/frabert/libdmusic/issues/17#issuecomment-416916532 "2018-08-29T11:16:10Z - Replied by Github Reply Comments"):
> Yeah RtAudio seems lazier that Portaudio, it doesn't like to convert between sampling rates / channel counts, but it is nicer to use from C++, I think it is staying for now

But isn't it the driver's/sound-adapter's job to do that? Or am I mistaken?

Anyway. It's a lot more difficult to test this way, if I have to constantly change sound adapter settings.

**Comment by frabert** [comment]
My guess is that it depends on what RtAudio uses as a back-end. As can be observed, WASAPI on Windows doesn't like to deal with mismatched settings, apparently. Maybe PulseAudio on Linux works, though, I have no way to try that out right now. PortAudio probably goes the extra mile to ensure even mismatched settings still work.

As for testing, I'm thinking of a way to render more than 2 channels easily. It's probably going to be just silence for every extra channel

**Comment by frabert** [comment]
dmplay should now be able to render to more than 2 channels, possibly with a (significant?) performance hit, dmrender instead refuses to do it.

**Comment by mzso** [comment]
Swell. Still works a bit oddly though. "dmplay -s 48000 AMB1.SGT" worked once for some reason, audio was playing. Then after the second try it didn't, I needed to add -c 8 to get sound again.

All in all it's still a fair bit of a pain having to match manually both the channel numbers and the sampling rate. For sampling I even get an error which specifies the rate I should use, so why can't it switch to it automatically? (There's nothing for channels so a user might get completely stumped to what the issue is.)

**Comment by mzso** [comment]
Oh wait, something else is wrong. With 8 channels it seems to get stuck on the first note which is drawn out indefinitely.

**Comment by frabert** [comment]
If that didn't work either, I'm afraid a rewrite of the synth engine is needed anyway, so this is going to be put on hold until a new synth engine is ready...
